Tray position like a module

raffael0 opened the pull request:
<!-- Please read our contributing guide before opening a PR: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md -->

## What type of PR is this? (check all applicable)

* [ ] Refactor
* [x] Feature
* [ ] Bug Fix
* [ ] Optimization
* [ ] Documentation Update
* [ ] Other: *Replace this with a description of the type of this PR*

## Description
This PR fixes the long-standing issue of only having 3 positioning options for the system tray. It does this by adding a new module which acts as placeholder. The actual tray window is then placed onto that tray module. Resizing of the tray and repositioning of the module is handled via signals.

These changes should also make it easier to completely break the tray handling out of the main bar.cpp into the new tray module.

## Related Issues & Documents
Fixes:  #1526
Fixes: #314

## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)
Since this PR adds a complete new option for tray positioning this has to be reflected in the wiki
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [ ] Does not require documentation changes

## TODO / Issues
* [ ] define how the tray positioning will be handled in the config. Will tray-position be deprecated?
* [x] figure out how to get the position where a tag will be placed in dispatch.cpp. The current approach only works on the left and in the center ([current implementation](https://github.com/raffael0/polybar/blob/e112604c94e47da0990e5d8c8d4387e8b03f18b9/src/tags/dispatch.cpp#L146))

I'd love some input on these issues

patrick96 commented on src/modules/tray.cpp:
```diff
@@ -0,0 +1,50 @@
+
+  tray_module::tray_module(const bar_settings& bar_settings, string name_)
+      : event_module<tray_module>(bar_settings, move(name_)) {
+    m_formatter->add(DEFAULT_FORMAT, "", {});
```
To be consistent, we should also use `format = <tray>` for the tray. This here just adds a pseudo format that doesn't really do anything (since it has no tags).

```suggestion
    m_formatter->add(DEFAULT_FORMAT, TAG_TRAY, {TAG_TRAY});
```

And `TAG_TRAY` could be defined as `<tray>`.

In that case, you also don't need `get_output` and can just check for `TAG_TRAY` in `tray_module::build`. See how the script module does this: https://github.com/polybar/polybar/blob/ab915fb724546e7abfcb35cbaf863ce2bca613ec/src/modules/script.cpp#L114-L122

patrick96 replied:
You can completely remove `get_output`. There is already a `get_output` defined for all modules in `base.inl` which takes care of invoking the `build` function.

patrick96 reviewed (changes requested):
Wow that's really cool :smiley: 

I appreciate you working on this. The tray has been a pain point for a long time and I have been dreading working on it forever.
This is surprisingly little code for what I thought would be a major undertaking.

Please let me know what your commitment for this PR is. Are you prepared to go through revisions until this is in a stable state? Or do you feel like this a "this is my code, take it or leave it" situation? Or something in between?
I don't want to push you until you lose interest, so it's good to know what kind of expectations you have here.
Of course feel free to at any point say that you no longer can/want to invest more time here.
We can just as well release this as an experimental, but mostly working, feature and improve it step by step.

---

While testing this, I found some issues, all of which happen using the following config:

```dosini
[bar/example]
modules-left = text tray text

enable-ipc = true

tray-position = adaptive

width = 50%
offset-x = 50%

[module/text]
type = custom/text
content = A

[module/tray]
type = internal/tray
```

* The `tray_manager` uses the position received from `tray_pos_change` as the absolute position for the tray window. However, the dispatcher only sends the x position relative to the bar window. This is an issue if you have two monitors and place the bar on the right one or when the bar has some `offset-x`.
* The tray module does not receive the `tray_width_change` signal and thus never updates its width and doesn't produce a proper offset tag. This causes modules that are placed after the tray to appear immediately after the previous module (and the tray overlaps them)
I'm pretty sure that's because the module is not attached to the signal handler and thus doesn't receive any signals.

I think these two are the most important points to get the module into a experimental state that we can release to users.

There are also some conceptual things and edge cases to consider:

* If you set `tray-position` to anything but `adaptive` but still have a tray module, the tray module will display the tray, but the bar will still cut off parts of the bar in `bar::parse`. We should enforce that if `tray-position = adaptive`, there must be exactly one tray module. And if it is anything else, no tray module must exist. We could also ditch `tray-position = adaptive` and just check for the existence of a tray module.
* If there is too much text on the bar, any excess text will just drop off the bar on the right. We have to decide (and properly handle) what happens when tray is shoved off the right side of the bar like that. My idea is that in that case the tray just stays at the right edge of the bar and never disappears. This can also easily be enforced by the `tray_manager` when it handles the `tray_pos_change` signal.
* We also have to think about how we want to move all the tray configurations that are currently into the bar section into the tray module. But this doesn't have priority at all yet, we can easily tackle this once the tray is stable.

These points are not yet important if we just release the tray module as an experimental feature, but we will need to address them to get to a stable state.

---

>  define how the tray positioning will be handled in the config. Will tray-position be deprecated?

In order to not break existing configs, we need to still support `tray-position`. Depending on whether we merge this as an experimental feature or try to get a stable implementation in this PR, we can already tell people to switch and that all the tray settings in the bar section will be deprecated. But leave it as-is for now.

> figure out how to get the position where a tag will be placed in dispatch.cpp. The current approach only works on the left and in the center

I think the solution here is to not immediately move the tray but only emit the tray position signal once the rendering is finished. Either the dispatch or the renderer need to keep track in which alignment and which relative position in that alignment the tray has to appear and at the end of the render cycle calculate the actual position.

raffael0 commented:
Hi, thanks for your quick and detailed response!

> Please let me know what your commitment for this PR is. Are you prepared to go through revisions until this is in a stable state? Or do you feel like this a "this is my code, take it or leave it" situation? Or something in between?

I'd ultimately like to see this become stable and I'm committed to continue working on it. Since I don't really understand some parts of the code works(modules, formatter etc.) I'll need some help in these areas and definitely appreciate any suggestions. The next few months will be quite busy for me so it may be that I'll need to take a break for a while but I'll definitely let you know before.  

> The tray_manager uses the position received from tray_pos_change as the absolute position for the tray window. However, the dispatcher only sends the x position relative to the bar window. This is an issue if you have two monitors and place the bar on the right one or when the bar has some offset-x

I also noticed this error and the tray positioning has been the most annoying issue I faced. Moving the tray position signal to the renderer sounds like a good idea. I never liked it in the dispatch. 
Am I correct that you propose moving it into renderer::end? 
Do I get the absolute and relative position like this?
In renderer:
```cpp 
 double absolute_x = m_rect.x + block_x(alignment);

```
in dispatch:
```cpp
double current_relative_x = renderer.get_x(*m_ctxt);
```

For now I'll look into the problem with the tray_width_change signal and incorporate your suggestions.

patrick96 commented:
> I'd ultimately like to see this become stable and I'm committed to continue working on it. Since I don't really understand some parts of the code works(modules, formatter etc.) I'll need some help in these areas and definitely appreciate any suggestions.

Just let me know if anything is unclear. Our code documentation isn't that good and a lot of code (especially the base module code) has been here for a long time (even before I started working on polybar) and its behavior isn't always obvious for me either.

> The next few months will be quite busy for me so it may be that I'll need to take a break for a while but I'll definitely let you know before.

That's completely fine :)
There is no rush to get this merged. I'd rather we take our time than do it sloppily.

---

The tray positioning is terrible. It's the main reason why I got stuck in my last big tray PR #1615.
The "proper" solution would be to make the tray window a child window of the bar window. Then we only have to position it relative to the parent and we also a avoid all the issues in #425.
However, that is probably quite a big change (bigger than what you did already), so I'm not sure if it's a good idea to try to force into this PR as well.
For now, let's keep the current system.

Unfortunately, neither of your proposed solutions work, both of them are still relative to the bar window. `m_rect` is relative to the bar window but only includes the inner area of the bar without the borders.

I think the following could work:
We store the tray position in the tag context (maybe as an `std::pair<alignment, int>`). The integer value is the x-position relative to the alignment. The dispatcher can easily find get that using `renderer.get_x`.

Once the dispatcher is finished, it calls into the renderer and the renderer can calculate the position relative to the inner area of the bar like this:

```cpp
double absolute_x = block_x(ctxt.tray.first) + ctxt.tray.second;
```

It then sends this to the tray_manager and the tray manager is responsible for calculating the absolute position. The tray gets passed the `bar_settings` which are necessary to get the absolute position in `setup`. However, I think it makes more sense to pass it into the constructor and also store it as a private field (`const bar_settings&`).
Because then, the tray manager can easily compute the absolute position when it receives the `tray_pos_change` signal like this:

```cpp
m_opts.orig_x = m_bar_opts.inner_area(true).x + evt.cast();
```

raffael0 commented:
- [x] Tray Width error: should be fixed
- [x] Tray positioning: Implemented you suggested fix. works with your example config
- [ ] Testing:  Since I added a new method to the `renderer`, I need to mock that in the [MockRenderer](https://github.com/raffael0/polybar/blob/185e271e835965c0198b2888c99b82591636c3e4/tests/unit_tests/tags/dispatch.cpp#L29) however my added Mock doesn't compile. Could you take a look at it?

patrick96 reviewed (changes requested):
This is coming along nicely :)

I have not had any problems when testing it out. The reviews are mainly cosmetic/architectural ones.

Please also run `clang-format` over all files that you have changed so that we have consistent formatting ([Style Guide](https://github.com/polybar/polybar/wiki/Style-Guide#code-formatting)).
You can run clang-format for all changed files using this command:

```
git diff --pretty="" --name-only "$(git merge-base HEAD upstream/master)..HEAD" | xargs clang-format -i --style=file
```

Note that this will completely mess up non-C++ files (cmake etc.), so just don't commit those ;)

After this round of reviews this should be in a state where we can merge this as an experimental feature. After that, we should think about the three optional points in my first review.

What do you prefer? Should we merge this as an experimental feature or first bring it into a stable state in this PR?

patrick96 commented:
This may also fix #651, but I haven't tested it yet.

raffael0 commented:
Hi,
just wanted to give an update on my progress. I'm on the break now I mentioned at the beginning so I currently don't have time to work on this. I'll  probably return next month. Maybe I can squeeze in the latest requested changes in the coming weeks to release it as an experimental feature.

patrick96 commented:
Thanks for letting me know :)

There is no need to rush here, take your time

Luekuu commented:
Hi Polygon Team, I'm new to the polygon/linux and github world. How i can "download" / "test" this tray-modul? 
Any beta files or so? 
Regards, Lukas

patrick96 commented:
@Luekuu This is very much not ready to use for end users. If you really want to, you can clone @raffael0's fork and [compile](https://github.com/polybar/polybar/wiki/Compiling) the [`working/variable-tray`](https://github.com/raffael0/polybar/tree/working/variable-tray) branch from source.

patrick96 commented:
Will review in depth later.

Your latest commit messed up the formatting in a bunch of files.
Please also revert the formatting changes in the cmake files.

patrick96 commented on doc/conf.py:
```diff
@@ -60,7 +61,7 @@ def get_version(root_path):
   # In all other builds conf.py is configured with cmake and put into the
   # build folder.
-  doc_path = '@doc_path@'
+  doc_path = '/home/raffael/Documents/Schule/21-22/ANDERE_PROJEKTE/polybar/doc'
```
Please revert all changes in this file

patrick96 commented on include/components/renderer_interface.hpp:
```diff
@@ -33,6 +33,8 @@ class renderer_interface {
    */
   virtual double get_alignment_start(const alignment align) const = 0;
 
+  virtual void apply_tray_position(unique_ptr <polybar::tags::context> &context) = 0;
```
```suggestion
  virtual void apply_tray_position(const tags::context& context) = 0;
```

It doesn't really make sense to pass a reference to a unique_ptr, passing a const-reference here is much cleaner (that's what the dispatcher does everywhere else anyway).

raffael0 commented:
Sorry about that. I ran clang-format but must have messed something up. I should have removed all of the wrong formatting now.

raffael0 commented:
Sorry about that. I have no idea how that happened. Should be fixed now.

patrick96 commented:
> Sorry about that. I have no idea how that happened. Should be fixed now.

Maybe your IDE auto-formatted the files.

patrick96 reviewed (changes requested):
Alright, looks good now :smiley: 

Things to do now to merge this as an experimental feature:

* [x] Add a changelog entry: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog
* [ ] Addressing the remaining reviews (mainly cosmetic)
* [ ] Remove the `bar_settings` argument from `tray_manager::setup`
* [ ] Resolve merge conflicts. Let me know if you need help with that.

After merging, we need to take these steps to make this feature stable

* [ ] Automatically hide the tray window when no tray position is set in the renderer
* [ ] Consider removing `tray-position = adaptive` again and just activate the tray when a tray module exists. The existence of a tray module should override any value for `tray-position`
* [ ] Fix the tray window being pushed off the bar. The tray window should always be fully within the bar window¨.
* [ ] Also support applicable tray settings in the tray module (`tray-maxsize`, `tray-background`, `tray-foreground`, `tray-padding`, `tray-scale`, and maybe `tray-offset-y`)
* [ ] Emit deprecation warnings for `tray-position` and all other tray settings in the bar section.$

These can be implemented step-by-step in separate PRs. I'll create an issue for them once we merge this.

raffael0 commented:
Hi. these commits should address all your points:
- Added changelog entry(I'm not sure about the wording here)
-  implemented the suggestions
- removed the bar_settings
- resolved the merge conflicts(I rebased my branch onto master. the tray new handling still works and the tests pass after rebasing)

patrick96 reviewed (changes requested):
Alright, last rounds of reviews :)

raffael0 commented on tests/unit_tests/tags/dispatch.cpp:
```diff
@@ -53,6 +54,10 @@ class FakeRenderer : public renderer_interface {
     return 0;
   };
 
+  void apply_tray_position(const polybar::tags::context& context) override {
```
thanks. Didn't know about that :)

patrick96 commented on CHANGELOG.md:
```diff
@@ -18,7 +18,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `internal/pulseaudio`: `reverse-scroll` option ([`#2664`](https://github.com/polybar/polybar/pull/2664))
 - `custom/script`: Repeat interval for script failure (`interval-fail`) and `exec-if` (`interval-if`) ([`#943`](https://github.com/polybar/polybar/issues/943), [`#2606`](https://github.com/polybar/polybar/issues/2606), [`#2630`](https://github.com/polybar/polybar/pull/2630))
-- `custom/text`: Loads the `format` setting, which supports the `<label>` tag, if the deprecated `content` is not defined ([`#1331`](https://github.com/polybar/polybar/issues/1331), [`#2673`](https://github.com/polybar/polybar/pull/2673), [`#2676`](https://github.com/polybar/polybar/pull/2676))
+- `custom/text`: Loads the `format` setting, which supports the `<label>` tag, if the deprecated `content` is not
+  defined ([`#1331`](https://github.com/polybar/polybar/issues/1331)
+  , [`#2673`](https://github.com/polybar/polybar/pull/2673), [`#2676`](https://github.com/polybar/polybar/pull/2676))
```
```suggestion
- `custom/text`: Loads the `format` setting, which supports the `<label>` tag, if the deprecated `content` is not defined ([`#1331`](https://github.com/polybar/polybar/issues/1331), [`#2673`](https://github.com/polybar/polybar/pull/2673), [`#2676`](https://github.com/polybar/polybar/pull/2676))
```

raffael0 commented:
Is this ready to be merged or am I missing an open issue?

patrick96 reviewed (approved):
> Is this ready to be merged or am I missing an open issue?

Everything's good, I was just busy.

patrick96 commented:
Awesome! Very happy to have this in polybar now :smiley: 

Thanks a lot @raffael0 for all your work!

I opened #2689 to track the next steps. If you want to work on some of that, feel free ;)

I'll be off for a week, so I won't be able to review anything.

raffael0 commented:
Great!
I'm also quite busy in coming weeks but I'll probably help out with the tasks still remaining

Vlek commented:
@raffael0, hello! I am trying to use this. However, there does not appear to be any documentation for it.

I have tried adding it to my bar and using `internal/tray`, but my polybar output gives an error saying that it is an unknown module. I made sure I'm using the latest version, 3.6.3 which appears to be after when this was merged to the master branch. Am I using this incorrectly? Is it not fully available yet?

raffael0 commented:
Hi,
this feature is targeted to be released in version 3.7. It was however merged into master and you can use it from there(I think polybar-git in the AUR should be on master or you can [build from source](https://github.com/polybar/polybar/wiki/Compiling)). We're still working on some issues to stabilize the module(See  #2689). Until then the feature is still experimental and the interface(eg. tray-position=adaptive) will probably change a bit until it's released.
